File: Deploy/GripperLoom.Examples/07-EventPublisher/CS/EventPublisher.cs
// ----------------------------------------------------------------
// * The LOOM.NET project http://rapier-loom.net *
// (C) Copyright by Wolfgang Schult. All rights reserved. 
// This code is licensed under the Apache License 2.0
// To get more information, go to http://loom.codeplex.com/license
// ----------------------------------------------------------------
using System;
using System.Collections.Generic;
using System.Text;

using Loom;
using Loom.JoinPoints;

namespace EventPublisher
{
    public delegate void ItemAdded(object sender, object item);
    
    /// <summary>
    /// 
    /// </summary>
    public interface IEventControl
    {
        event ItemAdded ItemAdded;
    }

    /// <summary>
    /// EventPublisher aspect class
    /// </summary>
    public class EventPublisher : AspectAttribute
    {
        [Introduce(typeof(IEventControl))] // introduces the IEventControl interface to the target class using the declared ItemAdded event
        public event ItemAdded ItemAdded;

        /// <summary>
        /// 
        /// </summary>
        /// <typeparam name="R">generic return type parameter used as wildcard</typeparam>
        /// <typeparam name="T">generic type parameter used as wildcard</typeparam>
        /// <param name="target">object representing the target object</param>
        /// <param name="item">type parameter of any type</param>
        /// <returns>wildcard - any type is matched except void</returns>
        [Call(Advice.Before)]
        [Include("Add")]
        public void Add<T>([JPTarget] object target, T item)
        {
            // triggers the ItemAdded event
            ItemAdded(target, item);
        }
    }
}

File: Deploy/RapierLoom.Examples/07-EventPublisher/VB/EventPublisher.vb
Public Class EventPublisher
    Inherits AspectAttribute

    <Introduce(GetType(IEventControl))> _
    Public Event ItemAdded As ItemAdded

    <Include("Add"), [Call](Advice.Before)> _
    Public Sub Add(Of T)(<JPTarget()> ByVal target As Object, ByVal item As T)
        RaiseEvent ItemAdded(target, item)
    End Sub

End Class

Public Interface IEventControl
    Event ItemAdded As ItemAdded
End Interface

Public Delegate Sub ItemAdded(ByVal sender As Object, ByVal item As Object)



